Webapp/sources/decorators.py

```diff
@@ -1,22 +1,139 @@
+    """
+    if search_str in request.args:
+        return request.args.get(search_str)
+    elif search_str in request.form.keys():
+        return request.form.get(search_str)
     else:
-        return workgroup
+        return input_value
+
+
+def _is_demo() -> bool:
+    """
+    checks whether demo or tutorial mode is active
+
+    Returns:
+        Bool: True if demo or tutorial is active else False
+
+    """
+    if request.form.get("demo") == "demo" or request.form.get("tutorial") == "tutorial":
+        return True
+    if request.args.get("demo") is None and request.args.get("tutorial") is None:
+        return False
+    return True
 
 
 def principal_investigator_required(f):
+    """
+        decorates function f to check whether user is a Principal Investigator in the workgroup,
+        and redirects to homepage if they are not.
+
+        The workgroup to search is passed either through a URL variable (eg: url/<workgroup>/...) or extracted from a
+        request using the _get_from_request function.
+
+        NOTE: redirect may fail if f returns json to front end
+
+        Returns:
+            redirect to home if user is not PI or SR else f
+        """
     @wraps(f)
-    def decorated_function(workgroup=None, *args, **kwargs):
-        workgroup = _get_workgroup(workgroup)
+    def decorated_function(*args, **kwargs):
+        workgroup = _get_from_request(kwargs.get("workgroup"), "workgroup")
         if services.workgroup.get_user_type(workgroup, current_user) != "principal_investigator":
             flash("You do not have permission to view this page")
             return redirect(url_for("main.index"))
-        return f(workgroup, *args, **kwargs)
+        kwargs["workgroup"] = workgroup
+        return f(*args, **kwargs)
+    return decorated_function
+
+
+def principal_investigator_or_senior_researcher_required(f):
+    """
+    decorates function f to check whether user is a Principal Investigator or Senior Researcher in the workgroup,
+    and redirects to homepage if they are not.
+
+    The workgroup to search is passed either through a URL variable (eg: url/<workgroup>/...) or extracted from a
+    request using the _get_from_request function.
+
+    NOTE: redirect may fail if f returns json to front end
+
+    Returns:
+        redirect to home if user is not PI or SR else f
+    """
+    @wraps(f)
+    def decorated_function(*args, **kwargs):
+        workgroup = _get_from_request(kwargs.get("workgroup"), "workgroup")
+        if services.workgroup.get_user_type(workgroup, current_user) not in ["principal_investigator", "senior_researcher"]:
+            flash("You do not have permission to view this page")
+            return redirect(url_for("main.index"))
+        kwargs["workgroup"] = workgroup
+        return f(*args, **kwargs)
+    return decorated_function
+
+
+def workgroup_member_required(f):
```

Comment:
Typehints for consistency

Reply:
I've added typehints. I think this is the correct way to do it but let me know if you have any other suggestions